File: UnrealAgent.Frontend/UI/Input/ImagePicker.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using UnrealAgent.Frontend.Infrastructure;

namespace UnrealAgent.Frontend.UI.Input;

public partial class ImagePicker : JsComponentBase
{
    // 이미지 변경 시 부모에 알리는 콜백
    [Parameter] public EventCallback OnChanged { get; set; }
    
    // 현재 첨부된 이미지의 MIME 타입
    public string? ImageMediaType { get; private set; }
    
    // 현재 첨부된 이미지의 Base64 데이터
    public string? ImageBase64 { get; private set; }
    
    // 파일 다이얼로그 참조
    private ElementReference FileInputRef;

    // .net에서 JS가 호출할 수 있는 참조
    private DotNetObjectReference<ImagePicker>? DotNetRef;

    protected override Task OnModuleLoaded()
    {
        DotNetRef = DotNetObjectReference.Create(this);
        return Task.CompletedTask;
    }
    
    // 파일 다이얼로그
    private async Task Pick()
    {
        await Module.InvokeVoidAsync("pick", FileInputRef, DotNetRef);
    }
    
    // JS에서 이미지 선택 완료시 호출된다.
    [JSInvokable]
    public async Task OnImagePicked(string MediaType, string Base64)
    {
        ImageMediaType = MediaType;
        ImageBase64 = Base64;
        
        await OnChanged.InvokeAsync();
    }
    
    // 첨부된 이미지를 제거
    public async Task Clear()
    {
        ImageMediaType = null;
        ImageBase64 = null;

        await OnChanged.InvokeAsync();
    }
    
    
    
    
    
    
    
    
}

File: UnrealAgent.Frontend/UI/Input/ImagePicker.razor.js

// 숨겨진 FIle input을 클릭하면 파일 다이얼로그를 연다
// 선택된 이미지를 Base64로 변환하여 C#에 전달
export function pick(FileInputRef, dotNetRef)
{
    // 파일 선택 시 호출되도록 함수 등록
    FileInputRef.onchange = function ()
    {
        var file = FileInputRef.files[0];
        if(!file)
            return;
        
        // 읽으면 호출되도록 함수 등록한다
        var reader = new FileReader();
        reader.onload = function()
        {
            var base64 = reader.result.split(",")[1];
            var mimeType = file.type || "image/png";
            dotNetRef.invokeMethodAsync("OnImagePicked", mimeType, base64);
        };
        
        // 파일 읽기 시작
        reader.readAsDataURL(file);
        
        FileInputRef.value = "";
    };
    
    // 파일 다이얼로그 활성화
    FileInputRef.click();
}